File: lib/tripFeed.js
// Server-side calendar feed for a planner. Resolves a planner by its UUID, loads
// its associated event dataset(s) to reconstruct tracked-session times, and renders
// the whole trip to an .ics using the SAME pure builders the client uses. Kept out
// of server.js so it can be unit-tested against fixtures.

import { readFile, readdir, stat } from 'fs/promises';
import { join, resolve, sep, basename } from 'path';
import {
  buildTripCalEvents,
  buildIcsCalendar,
  scheduleSessionsToCalEvents,
} from '../app/js/modules/plannerCalendar.js';

// Same session id the client derives (planner.js loadOneSchedule / makeSessionId):
// `${startTime}-${location}-${title}` slugified. Must match so trackedSessions
// (which store only sessionId) resolve to the right session.
const makeSessionId = (s) =>
  `${s.startTime}-${s.location}-${s.title}`.replace(/[^a-zA-Z0-9-]/g, '-');

function safeJoin(base, sub) {
  const full = resolve(join(base, sub));
  return full === base || full.startsWith(base + sep) ? full : null;
}
async function exists(p) {
  try {
    await stat(p);
    return true;
  } catch {
    return false;
  }
}

// A planner stores the nested dataset path (e.g. "events/ddd/2025-leuven.json").
// Older planners saved a flattened name — map those back via the catalog, mirroring
// the client's resolveEventFile.
async function resolveDatasetPath(file, dataDir) {
  const direct = safeJoin(dataDir, file);
  if (direct && (await exists(direct))) return direct;
  try {
    const catalog = JSON.parse(await readFile(join(dataDir, 'catalog.json'), 'utf8'));
    const files = (catalog.events || []).map((e) => e.file).filter(Boolean);
    const flatten = (f) => f.replace(/^events\//, '').replace(/\//g, '-');
    const match = files.find((f) => flatten(f) === file);
    if (match) {
      const p = safeJoin(dataDir, match);
      if (p && (await exists(p))) return p;
    }
  } catch {
    /* no catalog / unreadable → give up on this file */
  }
  return null;
}

// Load every associated event's sessions (id-tagged) + the primary event timezone.
async function loadEventData(eventFiles, dataDir) {
  let timezone = '';
  const sessions = [];
  for (let i = 0; i < eventFiles.length; i++) {
    const path = await resolveDatasetPath(eventFiles[i], dataDir);
    if (!path) continue;
    try {
      const data = JSON.parse(await readFile(path, 'utf8'));
      if (!timezone) timezone = data.event?.timezone || '';
      (data.items || []).forEach((s) => sessions.push({ ...s, id: makeSessionId(s) }));
    } catch {
      /* skip a corrupt dataset, keep the rest */
    }
  }
  return { sessions, timezone };
}

// Whole-trip .ics for a loaded planner object. Pure aside from the dataset reads.
export async function buildPlannerFeedIcs(planner, { dataDir } = {}) {
  const eventFiles =
    planner?._eventFiles?.length > 0
      ? planner._eventFiles
      : planner?._eventFile
        ? [planner._eventFile]
        : [];
  const { sessions, timezone } = await loadEventData(eventFiles, dataDir);
  // redactRefs: keep booking references (PNRs / confirmation numbers) OUT of the
  // public feed — a leaked link exposes the schedule, never booking credentials.
  const events = buildTripCalEvents(planner, {
    allSessions: sessions,
    timezone,
    include: {},
    redactRefs: true,
  });
  const calName = String(planner?._displayName || 'My trip').trim() || 'My trip';
  // 6h refresh hint — a change lands in the attendee's calendar within hours
  // (the in-app Now/Next covers the live minute-by-minute layer).
  return buildIcsCalendar(events, { calName, uidFor: (ev) => ev.uid, refreshMinutes: 360 });
}

// Public calendar feed of an ENTIRE event's programme (every session), for the
// schedule page's "subscribe to all sessions". No token — it's the same public
// data the schedule shows. `eventParam` is whitelisted against the catalog so it
// can't be used to read arbitrary files. Returns { ics, path } or null (→ 404).
export async function buildEventScheduleIcs(eventParam, { dataDir } = {}) {
  if (!eventParam) return null;
  let catalog;
  try {
    catalog = JSON.parse(await readFile(join(dataDir, 'catalog.json'), 'utf8'));
  } catch {
    return null;
  }
  const files = (catalog.events || []).map((e) => e.file).filter(Boolean);
  const flatten = (f) => f.replace(/^events\//, '').replace(/\//g, '-');
  // Whitelist: the request must name a known catalog dataset (nested or flattened).
  const match = files.find((f) => f === eventParam || flatten(f) === eventParam);
  if (!match) return null;
  const path = safeJoin(dataDir, match);
  if (!path) return null;
  let data;
  try {
    data = JSON.parse(await readFile(path, 'utf8'));
  } catch {
    return null;
  }
  const meta = data.event || {};
  const tz = meta.timezone || '';
  // Event map location: the dataset's own coords, else the geocode cache (see
  // scripts/geocode-events.mjs) — so a session gets a real map pin, not a blank.
  let geocache = {};
  try {
    geocache = JSON.parse(await readFile(join(dataDir, 'geocache.json'), 'utf8'));
  } catch {
    /* no geocache → sessions just carry text locations */
  }
  const place = [meta.venue, meta.location].filter((x) => x && String(x).trim()).join(', ');
  const evCoords = (() => {
    if (Number.isFinite(meta.latitude) && Number.isFinite(meta.longitude))
      return { lat: meta.latitude, lon: meta.longitude };
    const g = geocache[String(meta.location || '').trim()];
    return g && Number.isFinite(g.lat) ? { lat: g.lat, lon: g.lon } : null;
  })();
  // Tag items with the same session id the client derives, then map via the SHARED
  // builder so the feed and the schedule-page download produce identical VEVENTs.
  const items = (data.items || []).map((s) => ({ ...s, id: makeSessionId(s) }));
  const events = scheduleSessionsToCalEvents(items, {
    timezone: tz,
    place,
    coords: evCoords,
    uidBase: match,
  });
  const calName =
    [meta.designation, meta.location, meta.year].filter(Boolean).join(' ') || 'Schedule';
  // 12h refresh — a published programme changes far less often than a live trip.
  const ics = buildIcsCalendar(events, { calName, uidFor: (ev) => ev.uid, refreshMinutes: 720 });
  return { ics, path };
}

// Yield every stored planner (root files + one level of per-user subdirs, for
// multi-user mode) so we can find one by its UUID.
async function* scanPlanners(plannerDir) {
  let entries;
  try {
    entries = await readdir(plannerDir, { withFileTypes: true });
  } catch {
    return;
  }
  for (const ent of entries) {
    if (ent.isFile() && ent.name.endsWith('.json')) {
      const path = join(plannerDir, ent.name);
      const planner = await readJsonSafe(path);
      if (planner) yield { planner, userId: null, file: ent.name, path };
    } else if (ent.isDirectory()) {
      let sub;
      try {
        sub = await readdir(join(plannerDir, ent.name), { withFileTypes: true });
      } catch {
        continue;
      }
      for (const f of sub) {
        if (f.isFile() && f.name.endsWith('.json')) {
          const path = join(plannerDir, ent.name, f.name);
          const planner = await readJsonSafe(path);
          if (planner) yield { planner, userId: ent.name, file: f.name, path };
        }
      }
    }
  }
}
async function readJsonSafe(path) {
  try {
    return JSON.parse(await readFile(path, 'utf8'));
  } catch {
    return null;
  }
}

// A short-TTL in-memory index (uuid → file path) so the directory scan runs at most
// ~once/minute regardless of request volume — an unauthenticated poller (valid OR
// bogus uuid) can't amplify into a full scan per hit.
let _idxCache = { at: 0, dir: '', map: null };
const IDX_TTL_MS = 60_000;
async function getIndex(plannerDir) {
  const warm =
    _idxCache.map && _idxCache.dir === plannerDir && Date.now() - _idxCache.at < IDX_TTL_MS;
  if (warm) return _idxCache.map;
  const map = new Map();
  for await (const rec of scanPlanners(plannerDir)) {
    if (rec.planner?._id) map.set(rec.planner._id, rec.path);
  }
  _idxCache = { at: Date.now(), dir: plannerDir, map };
  return map;
}

// Find the planner whose `_id` matches `uuid`. The matched file is re-read fresh so
// token rotation / edits take effect immediately; only the scan is cached. A newly
// created planner becomes resolvable within the index TTL (≤ 60s).
export async function findPlannerById(uuid, { plannerDir } = {}) {
  if (!uuid) return null;
  const map = await getIndex(plannerDir);
  const path = map.get(uuid);
  if (!path) return null; // unknown id → answered from the cached index, no rescan
  const planner = await readJsonSafe(path);
  if (planner?._id !== uuid) {
    _idxCache = { at: 0, dir: '', map: null }; // stale mapping → rebuild next call
    return null;
  }
  return { planner, path, file: basename(path) };
}

File: tools/server/internal/planner/tripfeed.go
package planner

// A Go port of the trip half of lib/tripFeed.js and the planner entity mappers
// in app/js/modules/plannerCalendar.js — the whole-trip .ics.
//
// This is a person's travel: flights, hotels, the sessions they are going to.
// It is served on a token-gated URL, which means the feed itself is a
// credential, and booking references are redacted from it for that reason — a
// leaked link exposes the schedule, never the PNR.

import (
	"os"
	"path/filepath"
	"server/internal/archive"
	"server/internal/js"
	"server/internal/paths"
	"strings"
)

// travelModeLabels are the labels legs are titled with. The leading emoji is
// stripped by LegToCalEvent, but it has to be here to be stripped — the label
// is the string the client shows, and the .ics title is derived from it.
var travelModeLabels = map[string]string{
	"flight":    "✈ Flight",
	"train":     "🚂 Train",
	"bus":       "🚌 Bus",
	"ferry":     "⛴ Ferry",
	"car":       "🚗 Transfer",
	"taxi":      "🚕 Taxi",
	"rideshare": "📱 Rideshare",
	"other":     "↔ Other",
}

// ItineraryItemToCalEvent maps an itinerary item or a hosted event — they share
// one shape: title, date, optional time, optional location. No time means
// all-day.
func ItineraryItemToCalEvent(item *js.Value, timezone string) *CalEvent {
	date := item.Get("date")
	if !js.Truthy(date) {
		return nil
	}
	title := js.Trim(paths.FirstNonEmpty(
		js.String(item.Get("title")), js.String(item.Get("name")), "Itinerary item"))
	clock := js.Trim(js.String(item.Get("time")))
	endClock := js.Trim(js.String(item.Get("endTime")))

	event := &CalEvent{
		Title:    title,
		Start:    WallTime{Date: js.String(date), Time: clock},
		AllDay:   clock == "",
		Location: js.Trim(js.String(item.Get("location"))),
		// The item's notes land in the .ics DESCRIPTION.
		Description: js.Trim(paths.FirstNonEmpty(
			js.String(item.Get("notes")), js.String(item.Get("description")))),
		Timezone: js.Trim(paths.FirstNonEmpty(js.String(item.Get("timezone")), timezone)),
	}
	// An explicit finishing time gives the event a real end; without one the
	// VEVENT builder falls back to a default duration.
	if clock != "" && endClock != "" {
		event.End = &WallTime{Date: js.String(date), Time: endClock}
	}
	return event
}

// LegToCalEvent maps a travel leg. Start is departure, end is arrival, falling
// back to the departure date when no arrival date is set.
//
// includeRefs controls whether booking references reach the event body: on for
// a deliberate local download, OFF for the public subscription feed.
func LegToCalEvent(leg *js.Value, direction string, includeRefs bool) *CalEvent {
	if !js.Truthy(leg.Get("date")) {
		return nil
	}

	mode := js.String(leg.Get("mode"))
	label, known := travelModeLabels[mode]
	if !known {
		label = paths.FirstNonEmpty(mode, "Travel")
	}
	// Strip the leading emoji and its space: "✈ Flight" becomes "Flight".
	label = js.Trim(stripLeadingToken(label))

	route := archive.JoinNonEmpty(" → ", js.String(leg.Get("from")), js.String(leg.Get("to")))
	title := label
	if route != "" {
		title = label + ": " + route
	}

	departTime := js.Trim(js.String(leg.Get("departTime")))
	arriveTime := js.Trim(js.String(leg.Get("arriveTime")))

	var descriptionParts []string
	if includeRefs && js.Truthy(leg.Get("ref")) {
		descriptionParts = append(descriptionParts, "Ref: "+js.String(leg.Get("ref")))
	}
	if includeRefs && js.Truthy(leg.Get("confirmation")) {
		descriptionParts = append(descriptionParts,
			"Confirmation: "+js.String(leg.Get("confirmation")))
	}
	if direction != "" {
		descriptionParts = append(descriptionParts, "Direction: "+direction)
	}

	event := &CalEvent{
		Title:       title,
		Start:       WallTime{Date: js.String(leg.Get("date")), Time: departTime},
		AllDay:      departTime == "",
		Location:    js.Trim(js.String(leg.Get("from"))),
		Description: strings.Join(descriptionParts, "\n"),
	}
	if departTime != "" {
		event.End = &WallTime{
			Date: paths.FirstNonEmpty(js.String(leg.Get("arriveDate")), js.String(leg.Get("date"))),
			Time: paths.FirstNonEmpty(arriveTime, departTime),
		}
	}
	return event
}

// stripLeadingToken is `.replace(/^\S+\s/, ”)` — the first run of non-space
// characters and the space after it, and only when both are there.
func stripLeadingToken(s string) string {
	for i, r := range s {
		if r == ' ' || r == '\t' || r == '\n' || r == '\r' || r == '\f' || r == '\v' {
			if i == 0 {
				return s // starts with a space: \S+ cannot match
			}
			return s[i+len(string(r)):]
		}
	}
	return s
}

// AccommodationToCalEvent maps a stay to a single multi-day all-day event
// spanning check-in to check-out.
func AccommodationToCalEvent(stay *js.Value) *CalEvent {
	checkIn := js.Trim(js.String(stay.Get("checkIn")))
	if checkIn == "" {
		return nil
	}
	checkOut := paths.FirstNonEmpty(js.Trim(js.String(stay.Get("checkOut"))), checkIn)
	name := js.Trim(paths.FirstNonEmpty(js.String(stay.Get("name")), "Accommodation"))
	return &CalEvent{
		Title:    "Stay: " + name,
		Start:    WallTime{Date: checkIn},
		End:      &WallTime{Date: checkOut},
		AllDay:   true,
		Location: js.Trim(js.String(stay.Get("location"))),
	}
}

// TripInclude toggles each category of the trip. All on by default.
type TripInclude struct{ Travel, Stays, Items, Hosted, Sessions bool }

// AllTripCategories is the default: everything.
func AllTripCategories() TripInclude {
	return TripInclude{Travel: true, Stays: true, Items: true, Hosted: true, Sessions: true}
}

// BuildTripCalEvents aggregates a personal trip into calendar entries — travel
// legs, stays, itinerary items, hosted events and tracked sessions.
//
// Each entry carries a stable UID (`<entity>@<planner-uuid>`) so a re-import or
// a subscription refresh UPDATES rather than duplicates.
func BuildTripCalEvents(planner *js.Value, allSessions map[string]*js.Value, timezone string,
	include TripInclude, redactRefs bool) []*CalEvent {

	personal := planner.Get("personal")
	if personal == nil {
		personal = js.Obj()
	}
	namespace := paths.FirstNonEmpty(planner.Get("_id").Str(), "planner")

	var out []*CalEvent
	add := func(event *CalEvent, uidBase string) {
		if event != nil {
			event.UID = uidBase + "@" + namespace
			out = append(out, event)
		}
	}
	// An id that is absent interpolates as the string "undefined", and two
	// entities without ids therefore share a UID. That is the existing
	// behaviour and it is already in people's calendars.
	idOf := func(entity *js.Value) string {
		if id := entity.Get("id"); id.Exists() {
			return js.String(id)
		}
		return "undefined"
	}

	if include.Travel {
		for _, leg := range personal.Get("outboundLegs").Items() {
			add(LegToCalEvent(leg, "Outbound", !redactRefs), "leg-"+idOf(leg))
		}
		for _, leg := range personal.Get("returnLegs").Items() {
			add(LegToCalEvent(leg, "Return", !redactRefs), "leg-"+idOf(leg))
		}
		if js.Truthy(personal.Get("showLocalTravel")) {
			for _, leg := range personal.Get("localLegs").Items() {
				add(LegToCalEvent(leg, "Local", !redactRefs), "leg-"+idOf(leg))
			}
		}
	}
	if include.Stays {
		for _, stay := range personal.Get("accommodations").Items() {
			add(AccommodationToCalEvent(stay), "accom-"+idOf(stay))
		}
	}
	if include.Items {
		for _, item := range personal.Get("itinerary").Items() {
			add(ItineraryItemToCalEvent(item, timezone), "item-"+idOf(item))
		}
	}
	if include.Hosted {
		for _, hosted := range personal.Get("hostedEvents").Items() {
			// A hosted event's `description` is carried in as `notes`, which is
			// the field the itinerary mapper reads first.
			merged := js.Obj()
			for _, key := range hosted.Keys() {
				merged.Set(key, hosted.Get(key))
			}
			merged.Set("notes", hosted.Get("description"))
			add(ItineraryItemToCalEvent(merged, timezone), "hosted-"+idOf(hosted))
		}
	}
	if include.Sessions {
		for _, tracked := range personal.Get("trackedSessions").Items() {
			sessionID := tracked.Get("sessionId").Str()
			if session, found := allSessions[sessionID]; found {
				add(SessionToCalEvent(session, timezone), "session-"+sessionID)
			}
		}
	}
	return out
}

// BuildPlannerFeedIcs renders a whole trip.
//
// redactRefs is not a parameter: this feed is always the public one, and a
// booking reference has no business in a URL somebody can forward.
func BuildPlannerFeedIcs(planner *js.Value, dataDir, dtStamp string) string {
	var eventFiles []string
	for _, file := range planner.Get("_eventFiles").Items() {
		if name := file.Str(); name != "" {
			eventFiles = append(eventFiles, name)
		}
	}
	if len(eventFiles) == 0 {
		if single := planner.Get("_eventFile").Str(); single != "" {
			eventFiles = []string{single}
		}
	}

	sessions, timezone := loadEventSessions(eventFiles, dataDir)
	events := BuildTripCalEvents(planner, sessions, timezone, AllTripCategories(), true)

	calName := js.Trim(planner.Get("_displayName").Str())
	if calName == "" {
		calName = "My trip"
	}
	// 6h refresh — a change lands in the attendee's calendar within hours; the
	// in-app Now/Next covers the live minute-by-minute layer.
	return BuildIcsCalendar(events, calName, dtStamp, 0, 360)
}

// loadEventSessions loads every associated event's sessions, keyed by the id the
// client derives, plus the first timezone found.
func loadEventSessions(eventFiles []string, dataDir string) (map[string]*js.Value, string) {
	sessions := map[string]*js.Value{}
	timezone := ""
	for _, file := range eventFiles {
		path := resolveDatasetPath(file, dataDir)
		if path == "" {
			continue
		}
		raw, err := os.ReadFile(path)
		if err != nil {
			continue
		}
		data, err := js.ParseJSON(raw)
		if err != nil {
			// Skip a corrupt dataset, keep the rest.
			continue
		}
		if timezone == "" {
			timezone = data.Get("event").Get("timezone").Str()
		}
		for _, item := range data.Get("items").Items() {
			id := MakeSessionID(item)
			if _, seen := sessions[id]; !seen {
				// `Array.find` returns the FIRST match, so a duplicate id must
				// not displace the session already holding it.
				sessions[id] = item
			}
		}
	}
	return sessions, timezone
}

// resolveDatasetPath finds a dataset a planner refers to.
//
// A planner stores the nested path ("events/ddd/2025-leuven.json"). Older
// planners saved a flattened name, so those are mapped back through the catalog
// — mirroring the client's resolveEventFile.
func resolveDatasetPath(file, dataDir string) string {
	if direct := paths.SafeJoin(dataDir, file); direct != "" {
		if _, err := os.Stat(direct); err == nil {
			return direct
		}
	}
	raw, err := os.ReadFile(filepath.Join(dataDir, "catalog.json"))
	if err != nil {
		return ""
	}
	catalog, err := js.ParseJSON(raw)
	if err != nil {
		return ""
	}
	for _, entry := range catalog.Get("events").Items() {
		known := entry.Get("file").Str()
		if known == "" {
			continue
		}
		flattened := strings.ReplaceAll(strings.TrimPrefix(known, "events/"), "/", "-")
		if flattened != file {
			continue
		}
		if path := paths.SafeJoin(dataDir, known); path != "" {
			if _, err := os.Stat(path); err == nil {
				return path
			}
		}
	}
	return ""
}

// FoundPlanner is a planner and where it was read from.
type FoundPlanner struct {
	Planner *js.Value
	Path    string
	File    string
}

// FindPlannerByID finds the planner whose `_id` matches, scanning the planner
// directory and one level of per-user subdirectories.
//
// Deliberately NOT cached here, unlike the Node version's 60-second index. That
// cache exists so an unauthenticated poller cannot amplify a request into a full
// directory scan; a cache in a component that does not yet own the route would
// be a second answer to that question, and the wrong place to decide it. When
// this backs the live route it needs one — with a test for the stale-mapping
// path, which the original handles by invalidating.
func FindPlannerByID(uuid, plannerDir string) *FoundPlanner {
	if uuid == "" {
		return nil
	}
	var candidates []string

	entries, err := os.ReadDir(plannerDir)
	if err != nil {
		return nil
	}
	for _, entry := range entries {
		if !entry.IsDir() && strings.HasSuffix(entry.Name(), ".json") {
			candidates = append(candidates, filepath.Join(plannerDir, entry.Name()))
			continue
		}
		if entry.IsDir() {
			sub, err := os.ReadDir(filepath.Join(plannerDir, entry.Name()))
			if err != nil {
				continue
			}
			for _, file := range sub {
				if !file.IsDir() && strings.HasSuffix(file.Name(), ".json") {
					candidates = append(candidates,
						filepath.Join(plannerDir, entry.Name(), file.Name()))
				}
			}
		}
	}

	for _, path := range candidates {
		raw, err := os.ReadFile(path)
		if err != nil {
			continue
		}
		planner, err := js.ParseJSON(raw)
		if err != nil {
			continue
		}
		if planner.Get("_id").Str() == uuid {
			return &FoundPlanner{Planner: planner, Path: path, File: filepath.Base(path)}
		}
	}
	return nil
}
